File: cpp/LeetCode/No990.cpp
/*
You are given an array of strings equations that represent relationships between variables 
where each string equations[i] is of length 4 and takes one of two different forms: "xi==yi" or "xi!=yi".
Here, xi and yi are lowercase letters (not necessarily different) that represent one-letter variable names.

Return true if it is possible to assign integers to variable names so as to satisfy all the given equations, or false otherwise.
*/
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include "Debug\UnionFind.h"
using namespace std;

class Solution {
public:
    bool equationsPossible(vector<string>& equations) {
        UnionFind uf(26);
        vector<string> notEqual;
        for(string equation : equations){
            if(equation[1]=='!'){
                notEqual.push_back(equation);
                continue;
            }
            else{
                uf.connect(equation[0]-'a', equation[3]-'a');
            }
        }
        for(string equation : notEqual){
            if(uf.find(equation[0]-'a')==uf.find(equation[3]-'a')) return false;
        }
        return true;
    }
};

int main(){
    Solution solution;
    vector<string> equations;

    equations = {"a==b","b!=a"};
    cout << solution.equationsPossible(equations) << endl;

    equations = {"b==a","a==b"};
    cout << solution.equationsPossible(equations) << endl;
    return 0;
}

File: python/LeetCode Quiz/No990.py
'''
You are given an array of strings equations that represent relationships between variables 
where each string equations[i] is of length 4 and takes one of two different forms: "xi==yi" or "xi!=yi".
Here, xi and yi are lowercase letters (not necessarily different) that represent one-letter variable names.

Return true if it is possible to assign integers to variable names so as to satisfy all the given equations, or false otherwise.
'''
from typing import List

class Solution:
    def equationsPossible(self, equations: List[str]) -> bool:
        lowercase = 'abcdefghijklmnopqrstuvwxyz'
        graph = dict()
        for c in lowercase:
            graph[c] = set()
        checkList = []
        for i, s in enumerate(equations):
            if s[1]=='!':
                if s[0]==s[3]:
                    return False
                else:
                    checkList.append(i)
            else:
                if s[0]!=s[3]:
                    graph[s[0]].add(s[3])
                    graph[s[3]].add(s[0])
        # all the same letter will be in the same group

        checkMap = dict()
        for c in lowercase:
            checkMap[c] = c
        def dfs(node, c):
            if checkMap[node]==node:
                checkMap[node] = c
                for nei in graph[node]:
                    dfs(nei, c)
        for c in lowercase:
            if checkMap[c]==c:
                dfs(c, c)
        # make a map for letter to replaced the min letter in the same group
                
        for i in checkList:
            s = equations[i]
            if checkMap[s[0]]==checkMap[s[3]]:
                return False
        return True

def main():
    solution = Solution()

    equations = ["b!=f","c!=e","f==f","d==f","b==f","a==f"]
    print(solution.equationsPossible(equations))

if __name__=='__main__':
    main()
